### DISCUSSION TITLE
[PyTorch fundamentals] Change in numpy array does not change the tensor.

### DISCUSSION THREAD

**Discussion by send2gopal** [mrdbourke/pytorch-deep-learning#766] [Q&A]
Hi,

First of all, thanks for the video for Pytorch. I am going through the Pytorch fundamentals section of the video. There you have shown the demo where change in numpy array does not impact the  tensor which is working as expected for me as well , But when I try changing numpy array using index it actually changed the tensor, which means its sharing same memory. 

Here the question is if I do `numpy_array = numpy_array + 1` it does not change the tensor that means its doesnot share the same memory but in below scenario why it is sharing the same memory ?

`array = np.arange(1,8)
tensor = torch.from_numpy(array)
tensor, tensor.dtype

array[0] = 100
array, tensor`

![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/14090712/4b568b82-f207-416b-a295-d9dc64bc7072)

**Comment by shisirkha** [comment]
Because it needs the NumPy array to process the tensor and it shares same memory when converting index to a value

**Comment by wiertmir** [comment]
`array` is just a name pointing to the location in memory (memory reference, pointer)
And to access/change to whatever it is pointing we use `.` (dot), `()` or `[]` operators

The `=` operator doesn't change anything in the memory, it's role is to make "array" to point to some other location

To translate what `array = array + 1` does:
1. Make a new object which is a copy of whatever `array` is pointing to
2. Add 1 to all its elements
3. Make `array` to point to the new object (instead of original one)

The video is wrong and it could cause serious problems if people think that those objects (array and tensor) don't share same memory.

The following code could help to illustrate the issue with video:
```python
# Change the value of array, what will this do to `tensor`?
array_copy = array + 1
array[0] = 100
array_copy[0] = 99
array_copy, array, tensor
```
